refactor: replace JVM constraints with types

mlutze opened the pull request.

magnus-madsen commented:
The point here is to increase regularity and make everything more principled, right?

mlutze commented:
> The point here is to increase regularity and make everything more principled, right?
> i.e. we should have not have these extra constraints because they are equality constraints which we already support.

That's right. It especially means that the constraint solver follows the rule "I know how to unify two types if I know their kind or if I know their top-level shape." This was not the case when we had `Apply(JvmConstructor, <other stuff>)`.

mlutze commented:
@magnus-madsen RFR

I'm open to suggestions for type names.

magnus-madsen commented:
Can you write a very short description of how to best read the code? Thanks

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala:
```diff
@@ -467,17 +428,23 @@ object ConstraintSolver {
+    // Java types
+    case (tpe, java@Type.JvmToType(_, _)) =>
+      for {
+        (t1, p1) <- TypeReduction.simplify(tpe, renv, loc)
```
Need to get rid of the simplify calls.

mlutze commented:
> Can you write a very short description of how to best read the code? Thanks

# `Type.scala`
I introduce two new types:
- `JvmMember`
  - This is a type that corresponds to a Java field, method, or constructor. Inside it has a `JvmTemplate` which is one of
    - `JvmField(tpe, name)
    - `JvmConstructor(clazz, tpes)
    - `JvmMethod(...)`
    - `JvmStaticMethod(...)`
  - Once all the type variables are resolved, this type can be reduced to `Type.Cst(TypeConstructor.JvmField)` etc.
- `JvmToType`
  - This type extracts the return type or field type from a `Type.Cst(TypeConstructor.JvmField)` etc. when reduced.
  - This takes the place of the old `MethodReturnType` and `FieldType` types.

# `ConstraintGen.scala`

When we unify, instead of creating new constraints, we instead use equality constraints. I added the type rules in comments. They are pretty straightforward if you're familiar with the old strategy.

# `ConstraintSolver.scala`

I moved simplification into the `resolveOne` function. The logic for handling the old Jvm constraints has been moved to `TypeReduction.scala`.

I also added `JvmUnification.scala`, which only makes substitutions when one side is a variable.

# `TypeReduction.scala`

This now contains some of the logic previously in `ConstraintSolver.scala`

magnus-madsen commented:
@JonathanStarup Could you help with a review?

magnus-madsen commented:
Does your design support later adding a `JvmToEffect` assuming different methods have different effects?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/unification/JvmUnification.scala:
```diff
@@ -0,0 +1,31 @@
+import ca.uwaterloo.flix.language.ast.shared.Scope
+import ca.uwaterloo.flix.language.ast.{RigidityEnv, Type}
+
+object JvmUnification {
```
Not sure what this class is about?

mlutze replied:
This is to fulfill the principle that every kind gets its own XyzUnification. Obviously not a lot happens here, but I think it's nice to delegate anyway.

magnus-madsen replied:
OK

magnus-madsen reviewed (commented):
I think the implementation is very sensible. Good job 👍 

Picking the right representation of types is key and I think that has been achieved here.

My comments are minor and just about names.

mlutze commented:
> Does your design support later adding a `JvmToEffect` assuming different methods have different effects?

I think that should be fine, just matching the style of `JvmToType`. There's a complication with how it interacts with Boolean unification, but that will be there in any case.

magnus-madsen commented:
> > Does your design support later adding a `JvmToEffect` assuming different methods have different effects?
> 
> I think that should be fine, just matching the style of `JvmToType`. There's a complication with how it interacts with Boolean unification, but that will be there in any case.

Right-- because effects could now be learned later. Hm..

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/ast/Type.scala:
```diff
@@ -516,6 +540,97 @@ object Type {
+      *
+      * If `f` returns `Err` for any call, this function returns `Err`.
+      */
+    def traverse[E](f: Type => Result[Type, E]): Result[JvmTemplate, E] = this match {
```
OK

magnus-madsen replied:
Maybe move it out of Type then?

mlutze commented on main/src/ca/uwaterloo/flix/language/ast/Type.scala:
```diff
@@ -516,6 +540,97 @@ object Type {
+  /**
+    * An enumeration of the unresolved Java types.
+    */
+  sealed trait JvmTemplate {
```
`JvmMember` for this one?

magnus-madsen replied:
I feel like JvmOp is most descriptive?

magnus-madsen replied:
But I can accept JvmMember

magnus-madsen commented:
Ping me when I should review again.

mlutze commented:
> Ping me when I should review again.

@magnus-madsen I've addressed all the comments

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:
```diff
@@ -93,6 +93,12 @@ object Scheme {
+        Type.JvmToType(visitType(tpe), loc)
+
+      case Type.UnresolvedJvmType(template, loc) =>
+        Type.UnresolvedJvmType(template.map(visitType), loc)
```
@mlutze Need to search for "template" :)

mlutze replied:
fixed!
